test/general/backend_inclusion.pass.cpp

```diff
@@ -0,0 +1,69 @@
+// -*- C++ -*-
+//===-- backend_inclusion.pass.cpp ----------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+
+#include "support/utils.h"
+
+// The following test suites verify that the unused backend is disabled.
+#if !_ONEDPL_PAR_BACKEND_SERIAL
+#   ifdef _ONEDPL_PARALLEL_BACKEND_SERIAL_H
+#        error The parallel serial backend is used while it should not (_ONEDPL_PAR_BACKEND_SERIAL==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_OPENMP_BACKEND) && !ONEDPL_USE_OPENMP_BACKEND
+#   ifdef _ONEDPL_PARALLEL_BACKEND_OMP_H
+#       error The parallel OpenMP backend is used while it should not (ONEDPL_USE_OPENMP_BACKEND==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_TBB_BACKEND) && !ONEDPL_USE_TBB_BACKEND
+#   ifdef _ONEDPL_PARALLEL_BACKEND_TBB_H
+#       error The parallel TBB backend is used while it should not (ONEDPL_USE_TBB_BACKEND==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_DPCPP_BACKEND) && !ONEDPL_USE_DPCPP_BACKEND
+#   ifdef _ONEDPL_parallel_backend_sycl_H
+#       error The parallel DPC++ backend is used while it should not (ONEDPL_USE_DPCPP_BACKEND==0)
+#   endif
+#   ifdef _ONEDPL_parallel_backend_sycl_fpga_H
+#       error The DPC++ backend for the FPGA is used while it should not (ONEDPL_USE_DPCPP_BACKEND==0)
+#   endif
+#endif
+
+#if !ONEDPL_FPGA_DEVICE

```

Comment:
> Why the check pattern for this macro is different than the rest above?

Because as it's currently implemented in the code `ONEDPL_FPGA_DEVICE` macro is independent one. It can be set without `ONEDPL_USE_DPCPP_BACKEND`.
This check ensures that if `ONEDPL_FPGA_DEVICE` is turned off we (independently on ONEDPL_USE_DPCPP_BACKEND macro) should not include FPGA backend.
On the other hand I would prefer to follow documentation here and fix the implementation with saying that if `ONEDPL_FPGA_DEVICE` is defined while `ONEDPL_USE_DPCPP_BACKEND` is disabled it's preprocessor-time error.

>What's the logic of first checking defined(macro) then checking it for 0 (!macro)?

The logic is to check the public API and that we don't include what we should not (according to the initial intent of this test). Current state of test doesn't check the default behavior of DPC++ backend macros. I think that it would not be something different from what we have in the product code (not in tests) and I suppose the default case should be covered by other tests with different option for test building.

Reply:
> Why the check pattern for this macro is different than the rest above? 
Because the logic of using macros `ONEDPL_USE_*_BACKEND` and `ONEDPL_FPGA_DEVICE` is different. A specific `*` backend should definitely be disabled (`_ONEDPL_PARALLEL_BACKEND _ * _ H` undefined) only if the` ONEDPL_USE _ * _ BACKEND` macros are defined to be zero. The behavior when  `ONEDPL_USE _ * _ BACKEND` macros are not defined is governed by the current backend enable logic. But in the case of the ONEDPL_FPGA_DEVICE macro, the `_ONEDPL_parallel_backend_sycl_fpga_H` guard must be undefined if ONEDPL_FPGA_DEVICE is undefined or set to zero.

Reply:
I understand why a separate test for the macro is added, that's not my question. The question is, why for other public macros both the explicit definition (via `defined`) and the value are part of the test criteria, while for this macro there is no explicit definition check.

Reply:
> But in the case of the ONEDPL_FPGA_DEVICE macro, the _ONEDPL_parallel_backend_sycl_fpga_H guard must be undefined if ONEDPL_FPGA_DEVICE is undefined or set to zero.

And when undefined it's implicitly zero. OK that makes sense, thanks - but some comments would be helpful.